fendermint/testing/materializer/src/docker/mod.rs

```diff
@@ -0,0 +1,482 @@
+// Copyright 2022-2024 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use anyhow::Context;
+use async_trait::async_trait;
+use bollard::Docker;
+use ethers::{
+    core::rand::{rngs::StdRng, SeedableRng},
+    types::H160,
+};
+use fendermint_vm_actor_interface::eam::EthAddress;
+use fendermint_vm_core::{chainid, Timestamp};
+use fendermint_vm_genesis::{
+    ipc::{GatewayParams, IpcParams},
+    Account, Actor, ActorMeta, Collateral, Genesis, SignerAddr, Validator, ValidatorKey,
+};
+use fvm_shared::{bigint::Zero, econ::TokenAmount, version::NetworkVersion};
+use ipc_api::subnet_id::SubnetID;
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::BTreeMap,
+    path::{Path, PathBuf},
+};
+use tendermint_rpc::Url;
+
+use crate::{
+    manifest::Balance,
+    materializer::{Materializer, NodeConfig, SubmitConfig, SubnetConfig},
+    materials::{
+        export_json, export_script, import_json, DefaultAccount, DefaultDeployment, DefaultGenesis,
+        DefaultSubnet, Materials,
+    },
+    NodeName, RelayerName, ResourceHash, ResourceName, SubnetName, TestnetName,
+};
+
+mod container;
+mod network;
+mod node;
+mod relayer;
+
+pub use network::DockerNetwork;
+pub use node::DockerNode;
+pub use relayer::DockerRelayer;
+
+const STATE_JSON_FILE_NAME: &str = "materializer-state.json";
+
+const DOCKER_ENTRY_SCRIPT: &str = include_str!("../../scripts/docker-entry.sh");
+const DOCKER_ENTRY_FILE_NAME: &str = "docker-entry.sh";
+
+const PORT_RANGE_START: u32 = 30000;
+const PORT_RANGE_SIZE: u32 = 100;
```

Comment:
Do we bind these ports to the host?

Reply:
Yes, that's how the test will be able to interact with the container. Accidentally it means all tests will share the same materializer state, so they nodes across different tests don't get the same ports.